chainstate/src/detail/chainstateref.rs

```diff
@@ -438,46 +443,179 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
+
+        // Check token name
+        if token_ticker.len() > TOKEN_MAX_TICKER_LEN
+            || token_ticker.is_empty()
+            || !String::from_utf8_lossy(token_ticker).is_ascii()
         {
-            // check duplicate transactions
-            let mut txs_ids = BTreeSet::new();
-            for tx in block.transactions() {
-                let tx_id = tx.get_id();
-                let already_in_tx_id = txs_ids.get(&tx_id);
-                match already_in_tx_id {
-                    Some(_) => {
-                        return Err(CheckBlockTransactionsError::DuplicatedTransactionInBlock(
-                            tx_id,
-                            block.get_id(),
-                        ))
-                    }
-                    None => txs_ids.insert(tx_id),
-                };
+            return Err(TokensError::IssueErrorIncorrectTicker(tx_id, block_id));
+        }
+
+        // Check amount
+        if amount_to_issue == &Amount::from_atoms(0) {
+            return Err(TokensError::IssueErrorIncorrectAmount(tx_id, block_id));
+        }
+
+        // Check decimals
+        if number_of_decimals > &TOKEN_MAX_DEC_COUNT {
+            return Err(TokensError::IssueErrorTooManyDecimals(tx_id, block_id));
+        }
+
+        // Check URI
+        if metadata_uri.len() > TOKEN_MAX_URI_LEN
+            || !String::from_utf8_lossy(metadata_uri).is_ascii()
+        {
+            return Err(TokensError::IssueErrorIncorrectMetadataURI(tx_id, block_id));
+        }
+        Ok(())
+    }
+
+    fn check_transfer_data(
+        &self,
+        block_id: Id<Block>,
+        tx: &Transaction,
+        _token_id: &TokenId,
+        amount: &Amount,
+    ) -> Result<(), TokensError> {
+        // Check amount
+        ensure!(
+            amount > &Amount::from_atoms(0),
+            TokensError::TransferZeroTokens(tx.get_id(), block_id,)
+        );
+
+        Ok(())
+    }
+
+    fn check_tokens(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+        for tx in block.transactions() {
+            // We can't issue any count of token types in one tx
+            let issuance_count = get_tokens_issuance_count(tx);
+            ensure!(
+                issuance_count <= TOKEN_MAX_ISSUANCE_ALLOWED,
+                CheckBlockTransactionsError::CheckTokensError(
+                    TokensError::MultipleTokenIssuanceInTransaction(tx.get_id(), block.get_id()),
+                )
+            );
+
+            // Check assets
+            tx.outputs()
+                .iter()
+                .filter_map(|output| match output.value() {
+                    OutputValue::Coin(_) => None,
+                    OutputValue::Asset(asset) => Some(asset),
+                })
+                .try_for_each(|asset| self.check_asset(asset, tx, block))
+                .map_err(CheckBlockTransactionsError::CheckTokensError)?;
+        }
+        Ok(())
+    }
+
+    fn check_burn_data(
```

Comment:
All these checks for tokens should go to a separate submodule. No need to mix them all up in the chainstateref

Reply:
Ok